### PR TITLE
CORTX-28894 : Update HA mini provisioning to create system health key…

### PR DESCRIPTION
**Pull request by akash2144**
…s for disk resources

Signed-off-by: Akash Dudhane <akash.a.dudhane@seagate.com>

# Problem Statement
- Update HA mini provisioning to create system health keys for disk resources

# Design
-  During HA mini provisioning (ha_setup), system health keys need to be created for disk resources;  and appropriate status of these keys needs to be set 

# Coding
-  [x] Coding conventions are followed and code is consistent

# Testing 
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [x] Testing was performed with RPM

# Review Checklist 
- [x] PR is self reviewed
- [x] JIRA number/GitHub Issue added to PR
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained
- [ ] Is there a change in filename/package/module or signature? [Y/N]:  N
- [ ] If yes for above point, is a notification sent to all other cortx components? [Y/N] NA
- [ ] Side effects on other features (deployment/upgrade)? [Y/N]: N
- [ ] Dependencies on other component(s)? [Y/N]: N
-     If yes for above point, post link to the corresponding PR.

# Review Checklist 
- [ ] Is perfline test run and the report with and without the changes updated in the PR? [Y/N]: 

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

# Test Details : 

With Latest code changes consul keys are :
```
[root@cortx-ha-headless-svc /]# consul kv get -http-addr=consul-server.default.svc.cluster.local:8500 --recurse "cortx>ha"   cortx>ha>v1:
cortx>ha>v1>action>disk>failed:["publish"]
cortx>ha>v1>action>disk>online:["publish"]
cortx>ha>v1>action>node>failed:["publish"]
cortx>ha>v1>action>node>online:["publish"]
cortx>ha>v1>cluster_cardinality:{"num_nodes": 6, "node_list": ["113037cdcd2443bda5b6b47acd312421", "9b235fdd12774ea48c71dd7835461277", "623a56a21ab840fea48fafae92771e1d", "39561d21ffc04a3398b9cbd2e26b32ff", "32c0076d1ce14e229634a5c23cdafc12", "2aab06e9b2104623b7e74da0f3fbdfd4"]}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>health:{"events": [{"event_timestamp": "1647865470.8977857", "created_timestamp": "1647865471", "status": "degraded", "specific_info": null}], "action": {"modified_timestamp": "1647865471", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>health:{"events": [{"event_timestamp": "1647865470.8977857", "created_timestamp": "1647865471", "status": "degraded", "specific_info": null}], "action": {"modified_timestamp": "1647865471", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>health:{"events": [{"event_timestamp": "1647865470.8977857", "created_timestamp": "1647865471", "status": "degraded", "specific_info": null}], "action": {"modified_timestamp": "1647865471", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>113037cdcd2443bda5b6b47acd312421>cvg>cvg-01>disk>/dev/sdc>health:{"events": [{"event_timestamp": "1647865519", "created_timestamp": "1647865519", "status": "unknown", "specific_info": {"cvg_id": "cvg-01"}}], "action": {"modified_timestamp": "1647865519", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>113037cdcd2443bda5b6b47acd312421>cvg>cvg-01>disk>/dev/sdd>health:{"events": [{"event_timestamp": "1647865519", "created_timestamp": "1647865519", "status": "unknown", "specific_info": {"cvg_id": "cvg-01"}}], "action": {"modified_timestamp": "1647865519", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>113037cdcd2443bda5b6b47acd312421>cvg>cvg-01>disk>/dev/sde>health:{"events": [{"event_timestamp": "1647865519", "created_timestamp": "1647865519", "status": "unknown", "specific_info": {"cvg_id": "cvg-01"}}], "action": {"modified_timestamp": "1647865519", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>113037cdcd2443bda5b6b47acd312421>cvg>cvg-01>health:{"events": [{"event_timestamp": "1647865519", "created_timestamp": "1647865519", "status": "unknown", "specific_info": null}], "action": {"modified_timestamp": "1647865519", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>113037cdcd2443bda5b6b47acd312421>cvg>cvg-02>disk>/dev/sdf>health:{"events": [{"event_timestamp": "1647865520", "created_timestamp": "1647865520", "status": "unknown", "specific_info": {"cvg_id": "cvg-02"}}], "action": {"modified_timestamp": "1647865520", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>113037cdcd2443bda5b6b47acd312421>cvg>cvg-02>disk>/dev/sdg>health:{"events": [{"event_timestamp": "1647865520", "created_timestamp": "1647865520", "status": "unknown", "specific_info": {"cvg_id": "cvg-02"}}], "action": {"modified_timestamp": "1647865520", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>113037cdcd2443bda5b6b47acd312421>cvg>cvg-02>disk>/dev/sdh>health:{"events": [{"event_timestamp": "1647865520", "created_timestamp": "1647865520", "status": "unknown", "specific_info": {"cvg_id": "cvg-02"}}], "action": {"modified_timestamp": "1647865520", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>113037cdcd2443bda5b6b47acd312421>cvg>cvg-02>health:{"events": [{"event_timestamp": "1647865519", "created_timestamp": "1647865519", "status": "unknown", "specific_info": null}], "action": {"modified_timestamp": "1647865519", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>113037cdcd2443bda5b6b47acd312421>health:{"events": [{"event_timestamp": "1647865520.8414278", "created_timestamp": "1647865521", "status": "starting", "specific_info": {"generation_id": "cortx-data-ssc-vm-g4-rhev4-0762-68dfc746cd-j2w6g", "pod_restart": 0}}, {"event_timestamp": "1647865517", "created_timestamp": "1647865517", "status": "unknown", "specific_info": null}], "action": {"modified_timestamp": "1647865521", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>2aab06e9b2104623b7e74da0f3fbdfd4>cvg>cvg-01>disk>/dev/sdc>health:{"events": [{"event_timestamp": "1647865521", "created_timestamp": "1647865521", "status": "unknown", "specific_info": {"cvg_id": "cvg-01"}}], "action": {"modified_timestamp": "1647865521", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>2aab06e9b2104623b7e74da0f3fbdfd4>cvg>cvg-01>disk>/dev/sdd>health:{"events": [{"event_timestamp": "1647865520", "created_timestamp": "1647865520", "status": "unknown", "specific_info": {"cvg_id": "cvg-01"}}], "action": {"modified_timestamp": "1647865520", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>2aab06e9b2104623b7e74da0f3fbdfd4>cvg>cvg-01>disk>/dev/sde>health:{"events": [{"event_timestamp": "1647865521", "created_timestamp": "1647865521", "status": "unknown", "specific_info": {"cvg_id": "cvg-01"}}], "action": {"modified_timestamp": "1647865521", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>2aab06e9b2104623b7e74da0f3fbdfd4>cvg>cvg-01>health:{"events": [{"event_timestamp": "1647865520", "created_timestamp": "1647865520", "status": "unknown", "specific_info": null}], "action": {"modified_timestamp": "1647865520", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>2aab06e9b2104623b7e74da0f3fbdfd4>cvg>cvg-02>disk>/dev/sdf>health:{"events": [{"event_timestamp": "1647865521", "created_timestamp": "1647865521", "status": "unknown", "specific_info": {"cvg_id": "cvg-02"}}], "action": {"modified_timestamp": "1647865521", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>2aab06e9b2104623b7e74da0f3fbdfd4>cvg>cvg-02>disk>/dev/sdg>health:{"events": [{"event_timestamp": "1647865521", "created_timestamp": "1647865521", "status": "unknown", "specific_info": {"cvg_id": "cvg-02"}}], "action": {"modified_timestamp": "1647865521", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>2aab06e9b2104623b7e74da0f3fbdfd4>cvg>cvg-02>disk>/dev/sdh>health:{"events": [{"event_timestamp": "1647865521", "created_timestamp": "1647865521", "status": "unknown", "specific_info": {"cvg_id": "cvg-02"}}], "action": {"modified_timestamp": "1647865521", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>2aab06e9b2104623b7e74da0f3fbdfd4>cvg>cvg-02>health:{"events": [{"event_timestamp": "1647865521", "created_timestamp": "1647865521", "status": "unknown", "specific_info": null}], "action": {"modified_timestamp": "1647865521", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>2aab06e9b2104623b7e74da0f3fbdfd4>health:{"events": [{"event_timestamp": "1647865520.7422156", "created_timestamp": "1647865520", "status": "starting", "specific_info": {"generation_id": "cortx-data-ssc-vm-g4-rhev4-0763-db76f5f4f-zfsts", "pod_restart": 0}}, {"event_timestamp": "1647865519", "created_timestamp": "1647865519", "status": "unknown", "specific_info": null}], "action": {"modified_timestamp": "1647865520", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>32c0076d1ce14e229634a5c23cdafc12>cvg>cvg-01>disk>/dev/sdc>health:{"events": [{"event_timestamp": "1647865520", "created_timestamp": "1647865520", "status": "unknown", "specific_info": {"cvg_id": "cvg-01"}}], "action": {"modified_timestamp": "1647865520", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>32c0076d1ce14e229634a5c23cdafc12>cvg>cvg-01>disk>/dev/sdd>health:{"events": [{"event_timestamp": "1647865520", "created_timestamp": "1647865520", "status": "unknown", "specific_info": {"cvg_id": "cvg-01"}}], "action": {"modified_timestamp": "1647865520", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>32c0076d1ce14e229634a5c23cdafc12>cvg>cvg-01>disk>/dev/sde>health:{"events": [{"event_timestamp": "1647865520", "created_timestamp": "1647865520", "status": "unknown", "specific_info": {"cvg_id": "cvg-01"}}], "action": {"modified_timestamp": "1647865520", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>32c0076d1ce14e229634a5c23cdafc12>cvg>cvg-01>health:{"events": [{"event_timestamp": "1647865520", "created_timestamp": "1647865520", "status": "unknown", "specific_info": null}], "action": {"modified_timestamp": "1647865520", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>32c0076d1ce14e229634a5c23cdafc12>cvg>cvg-02>disk>/dev/sdf>health:{"events": [{"event_timestamp": "1647865520", "created_timestamp": "1647865520", "status": "unknown", "specific_info": {"cvg_id": "cvg-02"}}], "action": {"modified_timestamp": "1647865520", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>32c0076d1ce14e229634a5c23cdafc12>cvg>cvg-02>disk>/dev/sdg>health:{"events": [{"event_timestamp": "1647865520", "created_timestamp": "1647865520", "status": "unknown", "specific_info": {"cvg_id": "cvg-02"}}], "action": {"modified_timestamp": "1647865520", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>32c0076d1ce14e229634a5c23cdafc12>cvg>cvg-02>disk>/dev/sdh>health:{"events": [{"event_timestamp": "1647865520", "created_timestamp": "1647865520", "status": "unknown", "specific_info": {"cvg_id": "cvg-02"}}], "action": {"modified_timestamp": "1647865520", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>32c0076d1ce14e229634a5c23cdafc12>cvg>cvg-02>health:{"events": [{"event_timestamp": "1647865520", "created_timestamp": "1647865520", "status": "unknown", "specific_info": null}], "action": {"modified_timestamp": "1647865520", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>32c0076d1ce14e229634a5c23cdafc12>health:{"events": [{"event_timestamp": "1647865521.0039294", "created_timestamp": "1647865522", "status": "starting", "specific_info": {"generation_id": "cortx-data-ssc-vm-g4-rhev4-0314-7f7d8bcc6c-kd7cx", "pod_restart": 0}}, {"event_timestamp": "1647865518", "created_timestamp": "1647865518", "status": "unknown", "specific_info": null}], "action": {"modified_timestamp": "1647865522", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>39561d21ffc04a3398b9cbd2e26b32ff>health:{"events": [{"event_timestamp": "1647865520.784368", "created_timestamp": "1647865521", "status": "starting", "specific_info": {"generation_id": "cortx-server-ssc-vm-g4-rhev4-0762-5786949454-9jzl2", "pod_restart": 0}}, {"event_timestamp": "1647865518", "created_timestamp": "1647865518", "status": "unknown", "specific_info": null}], "action": {"modified_timestamp": "1647865521", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>623a56a21ab840fea48fafae92771e1d>health:{"events": [{"event_timestamp": "1647865521.0802948", "created_timestamp": "1647865522", "status": "starting", "specific_info": {"generation_id": "cortx-server-ssc-vm-g4-rhev4-0314-5c54f98d4-fv82d", "pod_restart": 0}}, {"event_timestamp": "1647865518", "created_timestamp": "1647865518", "status": "unknown", "specific_info": null}], "action": {"modified_timestamp": "1647865522", "status": "pending"}, "attributes": {}}
cortx>ha>v1>cortx>ha>system>cluster>118196a1d1094388be697ba3db13e367>site>NOT_DEFINED>rack>NOT_DEFINED>node>9b235fdd12774ea48c71dd7835461277>health:{"events": [{"event_timestamp": "1647865520.8792374", "created_timestamp": "1647865521", "status": "starting", "specific_info": {"generation_id": "cortx-server-ssc-vm-g4-rhev4-0763-68c947c544-w6cth", "pod_restart": 0}}, {"event_timestamp": "1647865518", "created_timestamp": "1647865518", "status": "unknown", "specific_info": null}], "action": {"modified_timestamp": "1647865521", "status": "pending"}, "attributes": {}}
cortx>ha>v1>events>disk>failed:["hare"]
cortx>ha>v1>events>disk>online:["hare"]
cortx>ha>v1>events>node>failed:["hare"]
cortx>ha>v1>events>node>online:["hare"]
cortx>ha>v1>events>subscribe>hare:["node>online", "node>failed", "disk>online", "disk>failed"]
cortx>ha>v1>message_type>hare:ha_event_hare

```

### CODE REVIEW

**Review by mukhtar-inamdar — COMMENTED** [review]
please check if your disk event will have ever specific_info i.e. event['payload']['specific_info'].
if ever it will have some object and not a generation_id then it will break in parser.
and generation_id is only for node event and not disk so it will fail if tomorrow if you have any object in 'specific_info' so let's change this also.
```
file: /root/cortx-ha/ha/core/event_analyzer/parser/parser.py
line: 191
if specific_info and specific_info["generation_id"] and source == HEALTH_EVENT_SOURCES.MONITOR.value:
```
*so you can make this check only for node events.*

**Review by mukhtar-inamdar — APPROVED** [review]

**Comment by akash2144** [comment]
> please check if your disk event will have ever specific_info i.e. event['payload']['specific_info']. if ever it will have some object and not a generation_id then it will break in parser. and generation_id is only for node event and not disk so it will fail if tomorrow if you have any object in 'specific_info' so let's change this also.
> 
> ```
> file: /root/cortx-ha/ha/core/event_analyzer/parser/parser.py
> line: 191
> if specific_info and specific_info["generation_id"] and source == HEALTH_EVENT_SOURCES.MONITOR.value:
> ```
> 
> _so you can make this check only for node events._

This is Check is added in PR#644. Currently we didn't receive any disk event from monitor.
https://github.com/Seagate/cortx-ha/pull/664/files#:~:text=if%20resource_type%20%3D%3D%20CLUSTER_ELEMENTS.NODE.value%3A

**Comment by shaileshsvc** [comment]
Please refresh the test output with latest code, this should reflect > as delimiter now.

**Comment by akash2144** [comment]
> Please refresh the test output with latest code, this should reflect > as delimiter now.

Done. Updated test results.

**Review by gauri-bhosale — CHANGES_REQUESTED** [review]
Can u please rebase branch with latest fault_k8s

**Comment by akash2144** [comment]
> Can u please rebase branch with latest fault_k8s

It's already rebased with latest fault_k8s. Please let me know if you see any warnings on PR.

**Review by shaileshsvc — APPROVED** [review]
Approved

**Review by gauri-bhosale — APPROVED** [review]

**Review by mariyappanp — APPROVED** [review]

**Review by Madhura-08 — APPROVED** [review]

File: ha/k8s_setup/ha_setup.py [hunk 1]
```diff
@@ -316,14 +318,32 @@ def _add_cvg_health(self) -> None:
...
 
-    def _add_health_event(self, node_id: str, resource_type: str, resource_id: str) -> None:
+    def _add_disk_health(self) -> None:
         """
-        Add health events for multiple resources (e.g. Node, CVG, disk)
+        Add disk health
+        """
+        _, nodes_list = self._confStoreAPI.get_cluster_cardinality()
+        for node in nodes_list:
+            cvg_list = ConftStoreSearch.get_cvg_list(self._index, node)
```
Comment (ArchanaLimaye) [inline ha/k8s_setup/ha_setup.py:327]: This logic is same as in add_cvg and code till here is getting executed twice, un-necessarily. 
We can have a new function to update disks for a particular cvg only and  add this call in _add_cvg_health thereby reusing the same code and reducing duplicate execution

**Comment by ArchanaLimaye** [comment]
@akash2144 from the test results,
-  looks like your setup had only 4 pods  ( 2 server , 2 data I guess) . Is this as expected ? There should be 3+3 right ?
- /dev/sda and /dev/sdb are missing. Is this expected as per cluster.conf

File: ha/k8s_setup/ha_setup.py (hunk 1 above)
Comment (akash2144) [reply ha/k8s_setup/ha_setup.py]: Done.

**Review by ArchanaLimaye — APPROVED** [review]

**Pull request merged**